jd4/case.py

```diff
@@ -256,20 +257,49 @@ def read_legacy_cases(config, open):
                           memory_bytes,
                           int(score_str))
 
-def read_yaml_cases(config, open):
-    for case in yaml.safe_load(config)['cases']:
-        if 'judge' not in case:
-            yield DefaultCase(partial(open, case['input']),
-                              partial(open, case['output']),
-                              parse_time_ns(case['time']),
-                              parse_memory_bytes(case['memory']),
-                              int(case['score']))
-        else:
-            yield CustomJudgeCase(partial(open, case['input']),
+def read_auto_cases(open, zip_file, time_limit='1s', memory_limit='128m'):
+    files = zip_file.namelist()
+    prefix = ''
+    cases = []
+    for i in files:
+        match = re_input.match(i)
+        if match:
+            prefix = match.group(1)
+            cases.append(int(match.group(2)))
+    if not len(cases):
+        raise FormatError('No test case found.')
+    cases.sort()
+    score = 100 // len(cases)
+    divider = len(cases) - 100 % len(cases)
+    for i, case in enumerate(cases):
+        yield DefaultCase(partial(open, prefix + str(case) + '.in'),
+                          partial(open, prefix + str(case) + '.out'),
+                          parse_time_ns(time_limit),
+                          parse_memory_bytes(memory_limit),
+                          score if i <= divider else score+1)
+
+def read_yaml_cases(config, open, zip_file):
+    cfg = yaml.safe_load(config)
+    if 'cases' not in cfg:
+        if 'time' not in cfg:
+            cfg['time'] = '1s'
```

Comment:
No `case` field means time/memory limit are same as `time` and `memory` field in the config file.

So `time` and `memory` should be required field, in case of typo made by user.

@twd2

Reply:
Agree.